keps/sig-storage/20200609-csi-service-account-token.md

```diff
@@ -0,0 +1,129 @@
+---
+title: Service Account Token for CSI Driver
+authors:
+  - "@zshihang"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-auth
+  - sig-storage
+reviewers:
+  - "@msau42"
+approvers:
+creation-date: 2020-06-09
+last-updated: 2020-06-09
+status: implementable
+---
+
+# Service Account Token for CSI Driver
+
+## Table of Contents
+
+<!-- toc -->
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [User stories](#user-stories)
+- [Proposal](#proposal)
+  - [CSIDriver Object](#csidriver-object)
+  - [Example Workflow](#example-workflow)
+- [Alternatives](#alternatives)
+- [Implementation History](#implementation-history)
+
+<!-- /toc -->
+
+## Summary
+
+This KEP proposes a way to obtain service account token for pods that the CSI
+drivers are provisioning volumes for. Since these tokens are valid only for a
+limited period, this KEP will also give the CSI drivers an option to remount
+volumes.
+
+## Motivation
+
+Currently, the only way that CSI drivers acquire service account tokens is to
+directly read the token in the file system. However, this approach has uncharming
+traits:
+
+1.  It will not work for csi drivers which run as a different non-root user than
+    the pods. See
+    [file permission section for service account token](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20180515-svcacct-token-volumes.md#file-permission).
+2.  CSI driver will have access to the secrets of pods that do not use it
+    because the CSI driver should have a `hostPath` volume for the `pods`
+    subdirectory to read the token.
+3.  The audience of the token is defaulted to kube apiserver.
+4.  The token is not guaranteed to be available (e.g.
+    `automountServiceAccountToken=false`).
+
+### User stories
+
+- HashiCorp Vault provider for secret store CSI driver requires service
+  account token of the pods they are mounting secrets at to authenticate to
+  Vaults. The provisioned secrets also have given TTL in Vault, so it is
+  neccessary get tokens after the inital mount.
+- Cert manager CSI dirver will create CertificateRequests on behalf of the
+  pods.
+- Amazon EFS CSI driver wants the service account tokens of pods to exchange for
+  AWS credentials.
+
+## Proposal
+
+Kubelet should inject audience-bounded service account tokens requested by the
+CSI driver from the `TokenRequest` API in `VolumeContext` to
+`NodePublishVolume`.
+
+### CSIDriver Object
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ... // existing fields
+
+    RequiresRemount *bool
+    ServiceAccountTokenAudience *string
```

Comment:
What are the valid values for this string?

Reply:
it is just an arbitrary non empty identifier.

the SP is supposed to send a [TokenReview](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#tokenreviewspec-v1-authentication-k8s-io) request with that audience included to validate the token.